pdns/dnsdist.cc

```diff
@@ -721,6 +723,48 @@ shared_ptr<DownstreamState> whashed(const NumberedServerVector& servers, const D
   return valrandom(dq->qname->hash(g_hashperturb), servers, dq);
 }
 
+/*
+ * @todo
+ * - test/benchmark other hashing methods
+ * - test/benchmark adding an avalanche algorithm on hashes
+ * @see https://github.com/haproxy/haproxy/blob/master/doc/internals/hashing.txt
+ */
+
+shared_ptr<DownstreamState> chashed(const NumberedServerVector& servers, const DNSQuestion* dq)
+{
+  std::map<unsigned int, shared_ptr<DownstreamState>> circle = {};
+  unsigned int qhash = dq->qname->hash(g_hashperturb);
+
+  for (const auto& d: servers) {
+    if (d.second->isUp()) {
+      if (d.second->hashes.empty()) {
+        // computes server's points
+        // @todo check if server's weight can change over time
+        auto w = d.second->weight;
+        while (w > 0) {
+          std::string uuid = boost::str(boost::format("%s-%d") % d.second->id % w);
+          unsigned int wshash = burtleCI((const unsigned char*)uuid.c_str(), uuid.size(), g_hashperturb);
+          d.second->hashes.insert(wshash);
+          --w;
+        }
+      }
+      for (const auto& h: d.second->hashes) {
+        // put server's hashes on the circle
+        circle.insert(std::make_pair(h, d.second));
```

Comment:
Do you think it would make sense to compute the circle when a server is added / removed or changes state?

Reply:
Hum, yes we could, but as the circle is pool dependent, it would require to change the signature of the policy handlers and pass the pool the query is assigned to. Didn't get the feeling it was an easy move as the policy system is quiet generic.

Reply:
I mean, to cache it, it would require to be aware of the context we are called in. Is that something we want ?

Reply:
Ah yes, it would lead to quite substantial changes so let's not go there until we have properly benchmarked that new policy.